geomesa-feature/src/main/scala/org/locationtech/geomesa/feature/serialization/AbstractReader.scala

```diff
@@ -0,0 +1,120 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the License);
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.locationtech.geomesa.feature.serialization
+
+import java.util.{Collections => JCollections, Map => JMap, UUID}
+
+import com.vividsolutions.jts.geom.Geometry
+import org.geotools.factory.Hints
+import org.locationtech.geomesa.feature.serialization.AbstractWriter.NULL_MARKER_STR
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes._
+
+/** Combines all readers.
+  * 
+  * Not thread safe.
+  */
+trait AbstractReader[Reader]
+  extends PrimitiveReader[Reader]
+  with NullableReader[Reader]
+  with CollectionReader[Reader]
+  with GeometryReader[Reader]
+  with HintKeyReader[Reader] {
+
+  /** Holds the serialization version number currently being read.  Must be managed externally. */
+  var version: Version = -1
+
+  def readUUID: DatumReader[Reader, UUID] = (in: Reader) =>  {
+    val mostSignificantBits = readLong(in)
+    val leastSignificantBits = readLong(in)
+    new UUID(mostSignificantBits, leastSignificantBits)
+  }
+
+  /** A [[DatumReader]] which reads a class name and then an object of that class.  If the class name is a null marker
+    * then ``null`` will be returned.
+    */
+  def readGeneric: DatumReader[Reader, AnyRef] = (reader) => {
+    val className = readString(reader)
+
+    if (className == NULL_MARKER_STR) {
+      null
+    } else {
+      val clazz = Class.forName(className)
+      selectReader(clazz).apply(reader)
+    }
+  }
+
+  /**
+   * A [[DatumReader]] for reading a map where the key and values may be any type.  The map may not be null. The reader
+   * will call ``readArrayStart(reader)`` and then, for each entry, read up to four items.
+   */
+  def readGenericMap: DatumReader[Reader, JMap[AnyRef, AnyRef]] = (reader) => {
+    var toRead = readArrayStart(reader)
+    val map = new java.util.HashMap[AnyRef, AnyRef](toRead)
+
+    while (toRead > 0) {
+      val key = readGeneric(reader)
+      val value = readGeneric(reader)
+
+      map.put(key, value)
+      toRead -= 1
+    }
+
+    map
+  }
+
+  /**
+   * @param clazz the [[Class]] of the object to be read
+   * @return a [[DatumReader]] capable of reading object of the given ``clazz``
+   */
+  def selectReader(clazz: Class[_], metadata: JMap[_ <: AnyRef, _ <: AnyRef] = JCollections.emptyMap(),
+                   isNullable: isNullableFn = notNullable): DatumReader[Reader, AnyRef] = {
+
+    val reader: DatumReader[Reader, AnyRef] = clazz match {
```

Comment:
this might be more readable as an if/else, since you're not unapplying or anything.